job-runner/webapp/server/classes/RunLimma.js

```diff
@@ -0,0 +1,201 @@
+  // create paths for files on the disk
+  var workDir = ntemp.mkdirSync("UpDownGenes");
+  console.log("workDir: ", workDir);
+
+  // create a sample group which is the combination of the two sample groups
+  // so that we can easily write out a file
+
+  var args = this.job.args;
+  var experimental =
+      SampleGroups.findOne(args.experimental_sample_group_id);
+  var reference =
+      SampleGroups.findOne(args.reference_sample_group_id);
+
+  // combine samples of same data set into single array
+  var dataSetHash = {};
+  _.each(experimental.data_sets.concat(reference.data_sets),
+      function (dataSet) {
+    // check if we've seen this data set already
+    var seenAlready = dataSetHash[dataSet.data_set_id];
+    if (!seenAlready) {
+      // if we haven't, set it up
+      seenAlready = {
+        data_set_name: dataSet.data_set_name,
+        sample_labels: [],
+      };
+    }
+
+    // combine the samples together
+    seenAlready.sample_labels =
+        seenAlready.sample_labels.concat(dataSet.sample_labels);
+    dataSetHash[dataSet.data_set_id] = seenAlready;
+  });
+  var comboSampleGroupDataSets = _.map(dataSetHash,
+      function (samplesAndName, data_set_id) {
+    return {
+      data_set_id: data_set_id,
+      data_set_name: samplesAndName.data_set_name,
+      sample_labels: samplesAndName.sample_labels,
+
+      // I think we can fake this
+      unfiltered_sample_count: 1,
+    };
+  });
+
+  var comboSampleGroupId = SampleGroups.insert({
+    name: "temp - created in RunLimma to call an adapter",
+    version: 1,
+    data_sets: comboSampleGroupDataSets,
+    value_type: experimental.value_type,
+
+    // invisible
+    collaborations: [],
+  });
+
+  // start the promise chain: woohoo!
+
+  var self = this;
+  var deferred = Q.defer();
+
+  // define the limma output files
+  var modelFitPath = path.join(workDir, "model_fit.tab");
+  var voomPlotPath = path.join(workDir, "mds.pdf");
+  var topGenePath = path.join(workDir, "top_genes.rnk");
+
+  // define this up here so we can pass data between the .thens
+  var limmaResultPass;
+
+  // write out the data for use in limma
+  Q.all([
+      spawnCommand(getSetting("genomic_expression_export"), [
+        "--sample_group_id", comboSampleGroupId,
+      ], workDir),
+
+      // phenotype file
+      spawnCommand(getSetting("limma_phenotype_export"), [
+        args.reference_sample_group_id,
+        args.experimental_sample_group_id
+      ], workDir),
+    ])
+    .spread(function (genomicExpSpawn, limmaPhenoSpawn) {
+      if (genomicExpSpawn.exitCode !== 0 || limmaPhenoSpawn.exitCode !== 0) {
+        throw "Problem writing files to disk.";
+      }
+
+      // run limma
+      return spawnCommand("Rscript", [
+        getSetting("limma_path"),
+        genomicExpSpawn.stdoutPath,
+        limmaPhenoSpawn.stdoutPath,
+        self.job.args.top_genes_count,
+
+        // "BH" or "none"
+        // TODO: make this an argument for the job?
```

Comment:
What is BH ? Is it fine that it's hardcoded in here? would we ever want to run it with "none", whatever that means?